precompiles/ics20/approve_common.go

```diff
@@ -0,0 +1,267 @@
+		}
+	}
+
+	// Only the origin can approve a transfer to the grantee address
+	expiration := ctx.BlockTime().Add(approvalExpiration).UTC()
+	if err := authzKeeper.SaveGrant(ctx, grantee.Bytes(), origin.Bytes(), transferAuthz, &expiration); err != nil {
+		return err
+	}
+
+	allocations := convertToAllocation(transferAuthz.Allocations)
+	// Emit the IBC transfer authorization event
+	return authorization.EmitIBCTransferAuthorizationEvent(
+		event,
+		ctx,
+		stateDB,
+		precompileAddr,
+		grantee,
+		origin,
+		allocations,
+	)
+}
+
+// Revoke implements the ICS20 Authorization revoke transactions.
+func Revoke(
+	ctx sdk.Context,
+	authzKeeper authzkeeper.Keeper,
+	precompileAddr, grantee, origin common.Address,
+	event abi.Event,
+	stateDB vm.StateDB,
+) error {
+	// NOTE: we do not need to check the expiration as it will return nil if both not found or expired
+	msgAuthz, _, err := authorization.CheckAuthzExists(ctx, authzKeeper, grantee, origin, TransferMsgURL)
+	if err != nil {
+		return fmt.Errorf(authorization.ErrAuthzDoesNotExistOrExpired, grantee, origin)
+	}
+
+	// check that the stored authorization matches the transfer authorization
+	if _, ok := msgAuthz.(*transfertypes.TransferAuthorization); !ok {
+		return authz.ErrUnknownAuthorizationType
+	}
+
+	if err = authzKeeper.DeleteGrant(ctx, grantee.Bytes(), origin.Bytes(), TransferMsgURL); err != nil {
+		return err
+	}
+
+	return authorization.EmitIBCTransferAuthorizationEvent(
+		event,
+		ctx,
+		stateDB,
+		precompileAddr,
+		grantee,
+		origin,
+		[]cmn.ICS20Allocation{},
+	)
+}
+
+// IncreaseAllowance implements the ICS20 Authorization increase allowance transactions.
+func IncreaseAllowance(
+	ctx sdk.Context,
+	authzKeeper authzkeeper.Keeper,
+	precompileAddr, grantee, origin common.Address,
+	sourcePort, sourceChannel, denom string,
+	amount *big.Int,
+	event abi.Event,
+	stateDB vm.StateDB,
+) error {
+	// NOTE: we do not need to check the expiration as it will return nil if both found or expired
+	msgAuthz, expiration, err := authorization.CheckAuthzExists(ctx, authzKeeper, grantee, origin, TransferMsgURL)
+	if err != nil {
+		return fmt.Errorf(authorization.ErrAuthzDoesNotExistOrExpired, grantee, origin)
+	}
+
+	// NOTE: we do not need to check the expiration as it will return nil if both found or expired
+	transferAuthz, ok := msgAuthz.(*transfertypes.TransferAuthorization)
+	if !ok {
+		return authz.ErrUnknownAuthorizationType
+	}
+
+	// Check if the allocations matches the arguments provided and returns the index of the allocation and coin found
+	spendLimit, allocationIdx, err := checkAllocationExists(transferAuthz.Allocations, sourcePort, sourceChannel, denom)
+	if err != nil {
+		return err
+	}
+
+	allowance := sdk.NewIntFromBigInt(amount)
+	if _, overflow := cmn.SafeAdd(spendLimit.Amount, allowance); overflow {
+		return errors.New(cmn.ErrIntegerOverflow)
```

Comment:
I think this error could provide more context where the integer overflow happened